### ISSUE TITLE
Filter/Sort Columns with Images

### ISSUE THREAD

**Issue by cmykkid** [Mottie/tablesorter#1102]
I'm sure TableSorter can do this, but I've been spinning my wheels for too long and need some help with the syntax.

I have some columns with KPI Health Indicators displayed. Red, Yellow, Green, and Grey. The images are part of a long URL and are named Red.png, Yellow.png, etc.

How do I have the Filter Select drop down display the choices "Red", "Yellow", etc and then filter on that choice?

Also, how do I control the sorting on these images? I.e. Have "Red" sort as top/bottom...not "Yellow".

Here's a starting point: http://jsfiddle.net/cmykkid/ays5obwt/12/

Thank you.

**Comment by Mottie** [comment]
Hi @cmykkid!

You can use the [`imgAttr`](http://mottie.github.io/tablesorter/docs/#imgattr) utility option to set the attribute of the image from where to obtain sortable data. It is set to `"alt"` by default.

```js
$(function(){
  $('table').tablesorter({
    // parse image title (value to be used while sorting & filtering)
    imgAttr : 'title',
    headers : {
      0 : { sorter: 'image' } // this parser is auto-detected, but will only work on the first image
    }
  });
});
```

The comment about it only working on the first image means that if there are two `img` inside the cell, then only the first will be targeted.

**Comment by cmykkid** [comment]
Thanks Mottie. Your reply appears to be targeting my second question regarding Sorting. 

But what about: I have some columns with KPI Health Indicators displayed. Red, Yellow, Green, and Grey. The images are part of a long URL and are named Red.png, Yellow.png, etc.

How do I have the Filter Select drop down display the choices "Red", "Yellow", etc and then filter on that choice?

**Comment by Mottie** [comment]
Oh, sorry, add a "filter-parsed" class name to the header ([ref](http://mottie.github.io/tablesorter/docs/example-widget-filter.html#classes)). This tells the filter to look at parsed data instead of any text contained in the cell.

**Comment by cmykkid** [comment]
I have the "filter-parsed" logic working, but it's setting the drop down filter choices to all Lower Case.
Try the Budget column. I'd like them to be Red, Yellow, etc...not red, yelllow, etc
Fiddle: http://jsfiddle.net/cmykkid/ays5obwt/16/
Thanks

**Comment by Mottie** [comment]
I think the easiest solution would be to add this css ([demo](http://jsfiddle.net/Mottie/ays5obwt/17/)): 

```css
select, option {
  text-transform: capitalize;
}
```

Or, if you want to change this internally - *your users will have to pay attention to case* - you would change both the [`ignoreCase`](http://mottie.github.io/tablesorter/docs/#ignorecase) and [`filter_ignoreCase`](http://mottie.github.io/tablesorter/docs/#widget-filter-ignorecase) options to `false` ([demo](http://jsfiddle.net/Mottie/ays5obwt/18/)).

**Comment by Mottie** [comment]
Also, I should mention that the `textExtraction` function in that demo is breaking the other columns.

If you don't use a parser, you can set a data-attribute on the cell which contains alternate text to be used for sorting and/or filtering. The default is `data-text`:

```html
<td data-text="Red">....</td>
```

This data-attribute can be changed using the [`textAttribute` option](http://mottie.github.io/tablesorter/docs/#textattribute).

**Comment by cmykkid** [comment]
Good catch. I didn't notice the other columns breaking. Ugh. I definitely don't want that.  I'm getting confused though.

If textExtraction is breaking the other columns then how do I get/use the data-text without textExtraction and parsing it?

Anyway you could adjust your demo to demonstrate this? I'd appreciate it.

**Comment by Mottie** [comment]
Well, because I didn't want to mess with your HTML, I fixed the `textExtraction` function ([demo](http://jsfiddle.net/Mottie/ays5obwt/19/)):

```js
// extract text from the table
textExtraction: {
	'.img' : function(node) {
		var $node = $(node);
		return $node.find('span[title]').attr('title') || $(node).text();
	}
}
```

And added these class names:

```html
<th class="img FilterProjStatus" data-value="">FilterProjStatus</th>
<th class="img filter-select filter-parsed">Budget</th>
```

**Comment by cmykkid** [comment]
Thanks Mottie. That would have taken me hours since I'm still green at this. More like days...

Just an FYI...if you're so inclined you're free to manipulate the HTML since I build the table dynamically after reading in a SharePoint list. If there's a more efficient way of coding this functionality than by all means. I can build the table how I like.

Also, I used to be able to type "red", for example, in the Schedule column filter and get a hit. Did this new functionality break that functionality? (At least I think I was able to.)

Thanks again.

**Comment by Mottie** [comment]
The `data-text="Red"` is actually added to each tbody td cell... 

```html
<td data-text="Red">img src=http://xyz.com/02010304/Grey.png</td>
```

Is there a reason why the image isn't really an image? The build-in image parser is the ideal solution - you don't need to do anything other than target the `img` title, if it has one.

As for searching, it's like I said before, setting `ignoreCase` and `filter_ignoreCase` to `true` will force the users to search for "Red" instead of "red". Capitalization matters. Which is why I said the css is the easier solution.

**Comment by cmykkid** [comment]
Finally getting back to this...

I have the KPI images displayed in my table and I'm able to get the column drop downs to display a user friendly name for the KPI images, but now I need the default sort to be set. See columns Budget and Schedule. Fiddle: http://jsfiddle.net/ays5obwt/23/

They are currently being sorted alphabetically based off the Title attribute, but I need them sorted in the order of Red, Yellow, Green, Gray.

The images are prefixed by their sorting order "1_Red", "2_Yellow", "3_Green", and "4_Gray" so I'm guessing I can use the number to get the desired sort.

Any help is appreciated.

**Comment by Mottie** [comment]
I updated [the demo](http://jsfiddle.net/ays5obwt/24/) as follows:

All titles are now set to a number plus the color:

```html
<td><img src=http://s17.postimg.org/5dyn3cf0b/1_Red_Status_Icon_X.png title="1_Red"></td>
```

Add `"filter-match"` and `"status"` class names to the header cells, so the filter only needs to match the title text instead of looking for an exact match (default for selects)

```html
<th class="filter-select filter-match status">Budget</th>
<th class="filter-select filter-match status">Schedule</th>
```

then add a `filter_selectSource` widget option:

```js
filter_selectSource: {
  '.status' : ['Red', 'Yellow', 'Green', 'Gray']
}
```

**Comment by cmykkid** [comment]
Thanks for the speedy reply Mottie. It works great.

I have one more issue. I would like all the table column filters to be reset when the user selects a different Project Status from the external select drop down menu, but not reset the Project Status external select drop down. This drop down is filtering on column 2 which is hidden.

I added "$table.trigger('filterReset');" to the drop down change logic, but the Project Status drop down is also being reset.

Is it possible to have all the table columns reset, but maintain filtering on the Project Status drop down? Here's a Fiddle: http://jsfiddle.net/ays5obwt/26/

Thank you.

**Comment by Mottie** [comment]
If you want the project status to override the other filters, then remove the `filter_external` option, and add this code to the `initialized` callback ([demo](http://jsfiddle.net/Mottie/ays5obwt/27/)):

```js
initialized: function () {
    $('select.search').on('change', function(){
        var query = [];
            query[$(this).data('column')] = $(this).val();
        $table.trigger('search', [query]);
    }).change();
}
```

**Comment by cmykkid** [comment]
At first I didn't know what this code change was doing. It was acting odd. Then I noticed the "data-column" for the Project Status select was switched to 1 instead of 2. Now it's working like a charm. (At least I was able to debug something on my own. :))

Here's a Fiddle: http://jsfiddle.net/ays5obwt/28/

I'll close this thread after a few days so I can review this. Thanks again for your help. Great tool.
